Python/Product/Debugger.Concord/Proxies/Structs/PyFrameObject311.cs

```diff
@@ -41,24 +43,46 @@ public PyFrameObject311(DkmProcess process, ulong address)
             CheckPyType<PyFrameObject311>();
         }
 
-        private PointerProxy<PyInterpreterFrame> f_frame {
+        public PointerProxy<PyInterpreterFrame> f_frame {
             get { return GetFieldProxy(_fields.f_frame); }
         }
 
         private PyInterpreterFrame GetFrame() {
             return f_frame.TryRead();
         }
 
-        public override PointerProxy<PyFrameObject> f_back => GetFieldProxy(_fields.f_back);
+        public override PointerProxy<PyFrameObject> f_back {
+            get {
+                // See the code here https://github.com/python/cpython/blob/a0866f4c81ecc057d4521e8e7a02f4e1fff175a1/Objects/frameobject.c#L1491,
+                // f_back can be null when there is actually a frame because it's created lazily.
+                var back = GetFieldProxy(_fields.f_back);
+                if (back.IsNull) {
+                    back = GetFrame().FindBackFrame();
+                }
+                return back;
+            }
+        }
 
         public override PointerProxy<PyCodeObject> f_code => GetFrame().f_code;
 
         public override PointerProxy<PyDictObject> f_globals => GetFrame().f_globals;
 
         public override PointerProxy<PyDictObject> f_locals => GetFrame().f_locals;
 
-        public override Int32Proxy f_lineno => GetFieldProxy(_fields.f_lineno);
-
         public override ArrayProxy<PointerProxy<PyObject>> f_localsplus => GetFrame().f_localsplus;
+
+        public override int ComputeLineNumber(DkmInspectionSession inspectionSession, DkmStackWalkFrame frame) {
+            var setLineNumber = GetFieldProxy(_fields.f_lineno).Read();
+            if (setLineNumber == 0) {
+                // We need to use the CppExpressionEvaluator to compute the line number from
+                // our frame object. This function here: https://github.com/python/cpython/blob/46710ca5f263936a2e36fa5d0f140cf9f50b2618/Objects/frameobject.c#L40-L41
+                var evaluator = new CppExpressionEvaluator(inspectionSession, 10, frame, DkmEvaluationFlags.TreatAsExpression);
+                var funcAddr = Process.GetPythonRuntimeInfo().DLLs.Python.GetFunctionAddress("PyFrame_GetLineNumber");
+                var frameAddr = Address; 
+                setLineNumber = evaluator.EvaluateInt32(string.Format("((int (*)(void *)){0})({1})", funcAddr, frameAddr), DkmEvaluationFlags.EnableExtendedSideEffects);
```

Comment:
This code is evaling a call to a function pointer. That looks like this : `((int (*)(void *))<func addr>)(<frame ptr addr>)`. It's how you call a function pointer in C/C++.

The function I'm calling is `PyFrame_GetLineNumber`, which if you look in the CPython source, it computes the line number if it's not already set. It's a public API, so it should always be there.